cast5/src/cast5.rs

```diff
@@ -0,0 +1,192 @@
+            .wrapping_add(S3[((i >> 8) & 0xff) as usize])
+            ^ S4[(i & 0xff) as usize]
+    }};
+}
+
+macro_rules! f3 {
+    ($D:expr, $m:expr, $r:expr) => {{
+        let i = ($m.wrapping_sub($D)).rotate_left(u32::from($r));
+        (S1[(i >> 24) as usize].wrapping_add(S2[((i >> 16) & 0xff) as usize])
+            ^ S3[((i >> 8) & 0xff) as usize])
+            .wrapping_sub(S4[(i & 0xff) as usize])
+    }};
+}
+
+impl BlockCipher for Cast5 {
+    type KeySize = U16;
+    type BlockSize = U8;
+    type ParBlocks = U1;
+
+    fn new(key: &GenericArray<u8, U16>) -> Self {
+        Self::new_varkey(&key).unwrap()
+    }
+
+    fn new_varkey(key: &[u8]) -> Result<Self, InvalidKeyLength> {
+        // Available key sizes are 40...128 bits.
+        if key.len() < 5 || key.len() > 16 {
+            return Err(InvalidKeyLength);
+        }
+        let mut cast5 = Cast5::init_state(key.len());
+
+        if key.len() < 16 {
+            // Pad keys that are less than 128 bits long.
+            let mut padded_key = [0u8; 16];
+            padded_key[..key.len()].copy_from_slice(key);
+            cast5.key_schedule(&padded_key[..]);
+        } else {
+            cast5.key_schedule(key);
+        }
+        Ok(cast5)
+    }
+
+    #[inline]
+    fn encrypt_block(&self, block: &mut Block) {
+        let masking = self.masking;
+        let rotate = self.rotate;
+
+        // (L0,R0) <-- (m1...m64). (Split the plaintext into left and
+        // right 32-bit halves L0 = m1...m32 and R0 = m33...m64.)
+        let l = BigEndian::read_u32(&block[0..4]);
+        let r = BigEndian::read_u32(&block[4..8]);
+        // (16 rounds) for i from 1 to 16, compute Li and Ri as follows:
+        //   Li = Ri-1;
+        //   Ri = Li-1 ^ f(Ri-1,Kmi,Kri), where f is defined in Section 2.2
+        // (f is of Type 1, Type 2, or Type 3, depending on i).
+        //
+        // Rounds 1, 4, 7, 10, 13, and 16 use f function Type 1.
+        // Rounds 2, 5, 8, 11, and 14 use f function Type 2.
+        // Rounds 3, 6, 9, 12, and 15 use f function Type 3.
+
+        let (l, r) = (r, l ^ f1!(r, masking[0], rotate[0]));
+        let (l, r) = (r, l ^ f2!(r, masking[1], rotate[1]));
+        let (l, r) = (r, l ^ f3!(r, masking[2], rotate[2]));
+        let (l, r) = (r, l ^ f1!(r, masking[3], rotate[3]));
+        let (l, r) = (r, l ^ f2!(r, masking[4], rotate[4]));
+        let (l, r) = (r, l ^ f3!(r, masking[5], rotate[5]));
+        let (l, r) = (r, l ^ f1!(r, masking[6], rotate[6]));
+        let (l, r) = (r, l ^ f2!(r, masking[7], rotate[7]));
+        let (l, r) = (r, l ^ f3!(r, masking[8], rotate[8]));
+        let (l, r) = (r, l ^ f1!(r, masking[9], rotate[9]));
+        let (l, r) = (r, l ^ f2!(r, masking[10], rotate[10]));
+        let (l, r) = (r, l ^ f3!(r, masking[11], rotate[11]));
+
+        let (l, r) = if self.small_key {
+            (l, r)
+        } else {
+            // Rounds 13..16 are only executed for keys > 80 bits.
```

Comment:
Hm, I wonder if it'll be worth to introduce two structs `Cast5` for 80-128 bit keys and `SmallCast5` for 40-80 bit keys. Have you measured the performance impact of adding this branching?

Reply:
There was no noticeable difference in the benchmarks I ran for 128bit keys befor and after